compiler/plugins/target/AMD-AIE/iree-amd-aie/Transforms/AMDAIELowerToAIE.cpp

```diff
@@ -1001,6 +1001,82 @@ class MoveAllDependenciesIntoOp : public OpRewritePattern<OpTy> {
   }
 };
 
+/// Utility to convert function arguments with memref element types' bitwidth
+/// either 8 or 16 to i32. This basically involves linearizing an N-d memref to
+/// 1-d i32 memref and adjusting the offset/size/stride metadata.
+static LogicalResult CastFunctionArgs(func::FuncOp funcOp,
+                                      PatternRewriter &rewriter) {
+  // only run on npu control functions
+  bool hasNpuOps = false;
+  funcOp.walk([&](AIEX::NpuDmaMemcpyNdOp dma) { hasNpuOps = true; });
+  if (!hasNpuOps) return failure();
+
+  // Cast all the function args to i32 types.
+  // This is in support of npu.dma_memcpy_nd which only allow 32bit integer
+  // types.
+  mlir::FunctionType funcType = funcOp.getFunctionType();
+  SmallVector<Type> argTypes(funcType.getInputs());
+  for (int i = 0, e = argTypes.size(); i < e; i++) {
+    auto memrefTy = dyn_cast<MemRefType>(argTypes[i]);
+    if (!memrefTy) continue;
+
+    unsigned int bitwidth = memrefTy.getElementTypeBitWidth();
+    if (bitwidth != 16 && bitwidth != 8) continue;
```

Comment:
Why specifically check for 16 and 8? Wouldn't this work for 4 bits as well?

Reply:
Yes, it should.

But since `AIRRtToNpu` kept this guarding for `8` and `16` bitwidth only, I didn't change - but we should be able to modify this gradually.

Reply:
The check is good for me like this, but maybe it's useful to add tests for 8 and 4 bits as well, so at least we're sure this logic works for different bitwidths? 8 bit will for sure come up soon and 4 bit might come up as well for BW reasons.